metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java

```diff
@@ -166,14 +166,37 @@ public void write(String sensorType, WriterConfiguration configurations, Iterabl
 
     }
 
-    BulkResponse resp = bulkRequest.execute().actionGet();
-
-    if (resp.hasFailures()) {
+    BulkResponse bulkResponse = bulkRequest.execute().actionGet();
 
-      throw new Exception(resp.buildFailureMessage());
+    return buildWriteReponse(tuples, bulkResponse);
+  }
 
+  protected BulkWriterResponse buildWriteReponse(Iterable<Tuple> tuples, BulkResponse bulkResponse) throws Exception {
+    // Elasticsearch responses are in the same order as the request, giving us an implicit mapping with Tuples
+    BulkWriterResponse writerResponse = new BulkWriterResponse();
+    if (bulkResponse.hasFailures()) {
+      Iterator<BulkItemResponse> respIter = bulkResponse.iterator();
+      Iterator<Tuple> tupleIter = tuples.iterator();
+      while (respIter.hasNext() && tupleIter.hasNext()) {
+        BulkItemResponse item = respIter.next();
+        Tuple tuple = tupleIter.next();
+
+        if (item.isFailed()) {
+          writerResponse.addError(item.getFailure().getCause(), tuple);
+        } else {
+          writerResponse.addSuccess(tuple);
+        }
+
+        // Should never happen, so fail the entire batch if it does.
+        if (respIter.hasNext() != tupleIter.hasNext()) {
+          throw new Exception(bulkResponse.buildFailureMessage());
```

Comment:
Good catch, I'll make that change.  IllegalStateException seems pretty reasonable.